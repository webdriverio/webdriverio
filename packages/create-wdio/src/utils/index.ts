
export { SUPPORTED_PACKAGE_MANAGERS } from '../constants.js'
