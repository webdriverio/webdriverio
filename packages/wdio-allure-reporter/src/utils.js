import process from 'process'
import CompoundError from './compoundError'
import { testStatuses, mochaEachHooks, mochaAllHooks, linkPlaceholder } from './constants'
import stripAnsi from 'strip-ansi'

/**
 * Get allure test status by TestStat object
 * @param test {Object} - TestStat object
 * @param config {Object} - wdio config object
 * @private
 */
export const getTestStatus = (test, config) => {
    if (config.framework === 'jasmine') {
        return testStatuses.FAILED
    }

    if (test.error.name && test.error.message) {
        const message = test.error.message.trim()
        return (test.error.name === 'AssertionError' || message.includes('Expect'))  ? testStatuses.FAILED : testStatuses.BROKEN
    }

    if (test.error.name) {
        return test.error.name === 'AssertionError' ? testStatuses.FAILED : testStatuses.BROKEN
    }

    if (test.error.stack) {
        const stackTrace = test.error.stack.trim()
        return (stackTrace.startsWith('AssertionError') || stackTrace.includes('Expect'))  ? testStatuses.FAILED : testStatuses.BROKEN
    }

    return testStatuses.BROKEN
}

/**
 * Check is object is empty
 * @param object {Object}
 * @private
 */
export const isEmpty = (object) => !object || Object.keys(object).length === 0

/**
 * Is mocha beforeEach / afterEach hook
 * @param title {String} - hook title
 * @returns {boolean}
 * @private
 */
export const isMochaEachHooks = title => mochaEachHooks.some(hook => title.includes(hook))

/**
 * Is mocha beforeAll / afterAll hook
 * @param title {String} - hook title
 * @returns {boolean}
 * @private
 */
export const isMochaAllHooks = title => mochaAllHooks.some(hook => title.includes(hook))

/**
 * Call reporter
 * @param {string} event  - event name
 * @param {Object} msg - event payload
 * @private
 */
export const tellReporter = (event, msg = {}) => {
    process.emit(event, msg)
}

/**
 * Properly format error from different test runners
 * @param {Object} test - TestStat object
 * @returns {Object} - error object
 * @private
 */
export const getErrorFromFailedTest = (test) => {
    if (test.errors && Array.isArray(test.errors)) {
        for(let i = 0; i < test.errors.length; i += 1){
            if(test.errors[i].message) test.errors[i].message = stripAnsi(test.errors[i].message)
            if(test.errors[i].stack) test.errors[i].stack = stripAnsi(test.errors[i].stack)
        }
        return test.errors.length === 1 ? test.errors[0] : new CompoundError(...test.errors)
    }
    if(test.error.message) test.error.message = stripAnsi(test.error.message)
    if(test.error.stack) test.error.stack = stripAnsi(test.error.stack)
    return test.error
}

/**
 * Substitute task id to link template
 * @param {string} template - link template
 * @param {string} id - task id
 * @returns {string} - link after substitution
 * @private
 */
export const getLinkByTemplate = (template, id) => {
    if (typeof template !== 'string') {
        return id
    }
    if(!template.includes(linkPlaceholder)) {
        throw Error(`The link template "${template}" must contain ${linkPlaceholder} substring.`)
    }
    return template.replace(linkPlaceholder, id)
}
