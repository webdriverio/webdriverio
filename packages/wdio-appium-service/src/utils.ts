import { basename, join, resolve } from 'path'
import { paramCase } from 'param-case'

import { ArgValue, KeyValueArgs } from './types'

const FILE_EXTENSION_REGEX = /\.[0-9a-z]+$/i

/**
 * Resolves the given path into a absolute path and appends the default filename as fallback when the provided path is a directory.
 * @param  {String} filePath         relative file or directory path
 * @param  {String} defaultFilename default file name when filePath is a directory
 * @return {String}                 absolute file path
 */
export function getFilePath (filePath: string, defaultFilename: string): string {
    let absolutePath = resolve(filePath)

    // test if we already have a file (e.g. selenium.txt, .log, log.txt, etc.)
    // NOTE: path.extname doesn't work to detect a file, cause dotfiles are reported by node to have no extension
    if (!FILE_EXTENSION_REGEX.test(basename(absolutePath))) {
        absolutePath = join(absolutePath, defaultFilename)
    }

    return absolutePath
}

export function formatCliArgs(args: KeyValueArgs | ArgValue[]): string[] {
    if (Array.isArray(args)) {
        return args.map(arg => sanitizeCliOptionValue(arg))
    }

    const cliArgs = []
    for (const key in args) {
        let value: ArgValue | ArgValue[] = args[key]
        // If the value is false or null the argument is discarded
        if ((typeof value === 'boolean' && !value) || value === null) {
            continue
        }

        cliArgs.push(`--${paramCase(key)}`)
        // Only non-boolean and non-null values are added as option values
        if (typeof value !== 'boolean') {
            cliArgs.push(sanitizeCliOptionValue(value))
        }
    }
    return cliArgs
}

export function sanitizeCliOptionValue (value: ArgValue) {
    const valueString = String(value)
    // Encapsulate the value string in single quotes if it contains a white space
    return /\s/.test(valueString) ? `'${valueString}'` : valueString
}

export function isWindows(): boolean {
    return process.platform === 'win32'
}
