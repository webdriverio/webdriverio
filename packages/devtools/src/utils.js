import fs from 'fs'
import path from 'path'
import { execFileSync } from 'child_process'
import logger from '@wdio/logger'
import { commandCallStructure, isValidParameter, getArgumentType, canAccess } from '@wdio/utils'
import { WebDriverProtocol } from '@wdio/protocols'

import cleanUp from './scripts/cleanUpSerializationSelector'
import { ELEMENT_KEY, SERIALIZE_PROPERTY, SERIALIZE_FLAG, ERROR_MESSAGES, PPTR_LOG_PREFIX } from './constants'

const log = logger('devtools')

export const validate = function (command, parameters, variables, ref, args) {
    const commandParams = [...variables.map((v) => Object.assign(v, {
        /**
         * url variables are:
         */
        required: true, // always required as they are part of the endpoint
        type: 'string'  // have to be always type of string
    })), ...parameters]

    const commandUsage = `${command}(${commandParams.map((p) => p.name).join(', ')})`
    const moreInfo = `\n\nFor more info see ${ref}\n`
    const body = {}

    /**
     * parameter check
     */
    const minAllowedParams = commandParams.filter((param) => param.required).length
    if (args.length < minAllowedParams || args.length > commandParams.length) {
        const parameterDescription = commandParams.length
            ? `\n\nProperty Description:\n${commandParams.map((p) => `  "${p.name}" (${p.type}): ${p.description}`).join('\n')}`
            : ''

        throw new Error(
            `Wrong parameters applied for ${command}\n` +
            `Usage: ${commandUsage}` +
            parameterDescription +
            moreInfo
        )
    }

    /**
     * parameter type check
     */
    for (const [i, arg] of Object.entries(args)) {
        const commandParam = commandParams[i]

        if (!isValidParameter(arg, commandParam.type)) {
            /**
             * ignore if argument is not required
             */
            if (typeof arg === 'undefined' && !commandParam.required) {
                continue
            }

            throw new Error(
                `Malformed type for "${commandParam.name}" parameter of command ${command}\n` +
                `Expected: ${commandParam.type}\n` +
                `Actual: ${getArgumentType(arg)}` +
                moreInfo
            )
        }

        /**
         * rest of args are part of body payload
         */
        body[commandParams[i].name] = arg
    }

    log.info('COMMAND', commandCallStructure(command, args))
    return body
}

export function getPrototype (commandWrapper) {
    const prototype = {}

    for (const [endpoint, methods] of Object.entries(WebDriverProtocol)) {
        for (const [method, commandData] of Object.entries(methods)) {
            prototype[commandData.command] = { value: commandWrapper(method, endpoint, commandData) }
        }
    }

    return prototype
}

export async function findElement (context, using, value) {
    /**
     * implicitly wait for the element if timeout is set
     */
    const implicitTimeout = this.timeouts.get('implicit')
    const waitForFn = using === 'xpath' ? context.waitForXPath : context.waitForSelector
    if (implicitTimeout && waitForFn) {
        await waitForFn.call(context, value, { timeout: implicitTimeout })
    }

    let element
    try {
        element = using === 'xpath'
            ? (await context.$x(value))[0]
            : await context.$(value)
    } catch (err) {
        /**
         * throw if method failed for other reasons
         */
        if (!err.message.includes('failed to find element')) {
            throw err
        }
    }

    if (!element) {
        return new Error(`Element with selector "${value}" not found`)
    }

    const elementId = this.elementStore.set(element)

    /**
     * return value has to be defined this way because of
     * https://github.com/microsoft/TypeScript/issues/37832
     */
    const returnValue = {}
    returnValue[ELEMENT_KEY] = elementId

    return returnValue
}

export async function findElements (context, using, value) {
    /**
     * implicitly wait for the element if timeout is set
     */
    const implicitTimeout = this.timeouts.get('implicit')
    const waitForFn = using === 'xpath' ? context.waitForXPath : context.waitForSelector
    if (implicitTimeout && waitForFn) {
        await waitForFn.call(context, value, { timeout: implicitTimeout })
    }

    const elements = using === 'xpath'
        ? await context.$x(value)
        : await context.$$(value)

    if (elements.length === 0) {
        return elements
    }

    return elements.map((element) => ({
        [ELEMENT_KEY]: this.elementStore.set(element)
    }))
}

/**
 * convert DevTools errors into WebDriver errors so tools upstream
 * can handle it in similar fashion (e.g. stale element)
 */
export function sanitizeError (err) {
    let errorMessage = err.message

    if (err.message.includes('Node is detached from document')) {
        err.name = ERROR_MESSAGES.staleElement.name
        errorMessage = ERROR_MESSAGES.staleElement.message
    }

    const stack = err.stack.split('\n')
    const asyncStack = stack.lastIndexOf('  -- ASYNC --')
    err.stack = errorMessage + '\n' + stack.slice(asyncStack + 1)
        .filter((line) => !line.includes('devtools/node_modules/puppeteer-core'))
        .join('\n')
    return err
}

/**
 * transform elements in argument list to Puppeteer element handles
 */
export async function transformExecuteArgs (args = []) {
    return Promise.all(args.map(async (arg) => {
        if (arg[ELEMENT_KEY]) {
            const elementHandle = await this.elementStore.get(arg[ELEMENT_KEY])

            if (!elementHandle) {
                throw getStaleElementError(arg[ELEMENT_KEY])
            }

            arg = elementHandle
        }

        return arg
    }))
}

/**
 * fetch marked elements from execute script
 */
export async function transformExecuteResult (page, result) {
    const isResultArray = Array.isArray(result)
    let tmpResult = isResultArray ? result : [result]

    if (tmpResult.find((r) => typeof r === 'string' && r.startsWith(SERIALIZE_FLAG))) {
        tmpResult = await Promise.all(tmpResult.map(async (r) => {
            if (typeof r === 'string' && r.startsWith(SERIALIZE_FLAG)) {
                return findElement.call(this, page, 'css selector', `[${SERIALIZE_PROPERTY}="${r}"]`)
            }

            return result
        }))

        await page.$$eval(`[${SERIALIZE_PROPERTY}]`, cleanUp, SERIALIZE_PROPERTY)
    }

    return isResultArray ? tmpResult : tmpResult[0]
}

export function getStaleElementError (elementId) {
    const error = new Error(
        `stale element reference: The element with reference ${elementId} is stale; either the ` +
        'element is no longer attached to the DOM, it is not in the current frame context, or the ' +
        'document has been refreshed'
    )
    error.name = 'stale element reference'
    return error
}

/**
 * Helper function to get a list of Puppeteer pages from a Chrome browser.
 * In case many headless browser are run in parallel there are situations
 * where there are no pages because the machine is busy booting the headless
 * browser.
 *
 * @param  {Puppeteer.Browser} browser  browser instance
 * @return {Puppeteer.Page[]}           list of browser pages
 */
export async function getPages (browser, retryInterval = 100) {
    const pages = await browser.pages()

    if (pages.length === 0) {
        log.info('no browser pages found, retrying...')

        /**
         * wait for some milliseconds to try again
         */
        await new Promise((resolve) => setTimeout(resolve, retryInterval))
        return getPages(browser)
    }

    return pages
}

export function sort(installations, priorities) {
    const defaultPriority = 10
    return installations
        // assign priorities
        .map((inst) => {
            for (const pair of priorities) {
                if (pair.regex.test(inst)) {
                    return { path: inst, weight: pair.weight }
                }
            }

            return { path: inst, weight: defaultPriority }
        })
        // sort based on priorities
        .sort((a, b) => (b.weight - a.weight))
        // remove priority flag
        .map(pair => pair.path)
}

/**
 * helper utitlity to clone a list
 * @param  {Any[]} arr  list of things
 * @return {Any[]}      new list of same things
 */
export function uniq(arr) {
    return Array.from(new Set(arr))
}

/**
 * Look for edge executables by using the which command
 */
export function findByWhich (executables, priorities) {
    const installations = []
    executables.forEach((executable) => {
        try {
            const browserPath = execFileSync(
                'which',
                [executable],
                { stdio: 'pipe' }
            ).toString().split(/\r?\n/)[0]

            if (canAccess(browserPath)) {
                installations.push(browserPath)
            }
        } catch (e) {
            // Not installed.
        }
    })

    return sort(uniq(installations.filter(Boolean)), priorities)
}

/**
 * monkey patch debug package to log CDP messages from Puppeteer
 */
export function patchDebug (scoppedLogger) {
    /**
     * log puppeteer messages
     */
    let puppeteerDebugPkg = path.resolve(
        path.dirname(require.resolve('puppeteer-core')),
        'node_modules',
        'debug')

    /**
     * check if Puppeteer has its own version of debug, if not use the
     * one that is installed for all packages
     */
    if (!fs.existsSync(puppeteerDebugPkg)) {
        /**
         * let's not get caught by our dep checker, therefor
         * define package name in variable first
         */
        const pkgName = 'debug'
        puppeteerDebugPkg = require.resolve(pkgName)
    }

    require(puppeteerDebugPkg).log = (msg) => {
        if (msg.includes('puppeteer:protocol')) {
            msg = msg.slice(msg.indexOf(PPTR_LOG_PREFIX) + PPTR_LOG_PREFIX.length).trim()
        }
        scoppedLogger.debug(msg)
    }
}
