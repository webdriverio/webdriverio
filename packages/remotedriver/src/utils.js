import logger from '@wdio/logger'

const log = logger('remotedriver')

import { isValidParameter, getArgumentType, commandCallStructure } from 'webdriver/src/utils'

export const validate = function (command, parameters, variables, ref, args) {
    const commandParams = [...variables.map((v) => Object.assign(v, {
        /**
         * url variables are:
         */
        required: true, // always required as they are part of the endpoint
        type: 'string'  // have to be always type of string
    })), ...parameters]

    const commandUsage = `${command}(${commandParams.map((p) => p.name).join(', ')})`
    const moreInfo = `\n\nFor more info see ${ref}\n`
    const body = {}

    /**
     * parameter check
     */
    const minAllowedParams = commandParams.filter((param) => param.required).length
    if (args.length < minAllowedParams || args.length > commandParams.length) {
        const parameterDescription = commandParams.length
            ? `\n\nProperty Description:\n${commandParams.map((p) => `  "${p.name}" (${p.type}): ${p.description}`).join('\n')}`
            : ''

        throw new Error(
            `Wrong parameters applied for ${command}\n` +
            `Usage: ${commandUsage}` +
            parameterDescription +
            moreInfo
        )
    }

    /**
     * parameter type check
     */
    for (const [i, arg] of Object.entries(args)) {
        const commandParam = commandParams[i]

        if (!isValidParameter(arg, commandParam.type)) {
            /**
             * ignore if argument is not required
             */
            if (typeof arg === 'undefined' && !commandParam.required) {
                continue
            }

            throw new Error(
                `Malformed type for "${commandParam.name}" parameter of command ${command}\n` +
                `Expected: ${commandParam.type}\n` +
                `Actual: ${getArgumentType(arg)}` +
                moreInfo
            )
        }

        /**
         * rest of args are part of body payload
         */
        body[commandParams[i].name] = arg
    }

    log.info('REMOTE COMMAND', commandCallStructure(command, args))
    return body
}
