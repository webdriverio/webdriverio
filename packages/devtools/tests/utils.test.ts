import path from 'path'
// not a real package but needed for testing
// @ts-ignore
import { log as pptrDebugLog } from 'pptrDebug'
import childProcess from 'child_process'
import { canAccess } from '@wdio/utils'

import {
    validate, getPrototype, findElement, findElements, getStaleElementError,
    sanitizeError, transformExecuteArgs, transformExecuteResult, getPages,
    uniq, findByWhich, patchDebug
} from '../src/utils'

const debug = jest.requireActual('debug')

const command = {
    endpoint: '/session/:sessionId/element/:elementId/element',
    method: 'POST',
    command: 'findElementFromElement',
    ref: 'https://w3c.github.io/webdriver/webdriver-spec.html#dfn-find-element-from-element',
    variables: [{
        'name': 'elementId',
        'description': 'the id of an element returned in a previous call to Find Element(s)'
    }],
    parameters: [{
        'name': 'using',
        'type': 'string',
        'description': 'a valid element location strategy',
        'required': true
    }, {
        'name': 'value',
        'type': 'string',
        'description': 'the actual selector that will be used to find an element',
        'required': true
    }, {
        'name': 'customParam',
        'type': 'number',
        'description': 'a random not required param',
        'required': false
    }]
}

let pageMock = {
    waitForSelector: jest.fn(),
    waitForXPath: jest.fn(),
    $$eval: jest.fn(),
    $$: jest.fn(),
    $x: jest.fn(),
    $: jest.fn()
}

jest.mock('fs', () => {
    return {
        existsSync: (pkgName: string) => pkgName === 'pptrDebug'
    }
})

jest.mock('child_process', () => {
    let returnValue = false
    return {
        execFileSync: jest.fn().mockImplementation(() => {
            if (!returnValue) {
                throw new Error('foo not found')
            }
            return returnValue
        }),
        shouldReturn: (value: any) => (returnValue = value)
    }
})

jest.mock('path', () => {
    let resolveResult = 'debug'
    const resolve = jest.fn(() => resolveResult)
    const setResolveResult = (result: any) => (resolveResult = result)
    return { resolve, setResolveResult, dirname: jest.fn() }
})

describe('validate', () => {
    it('should fail if wrong arguments are passed in', () => {
        expect(() => validate(command.command, command.parameters, command.variables, command.ref, []))
            .toThrow(/Wrong parameters applied for findElementFromElement/)
    })

    it('should fail if arguments are malformed', () => {
        expect(() => validate(command.command, command.parameters, command.variables, command.ref, ['123', 123, '123']))
            .toThrow(/Malformed type for "using" parameter of command/)
    })

    it('should fail if not required param has wrong type', () => {
        expect(() => validate(command.command, command.parameters, command.variables, command.ref, ['123', '123', '123', 'foobar']))
            .toThrow(/Malformed type for "customParam" parameter of command/)
    })

    it('should do a proper request', () => {
        const body = validate(command.command, command.parameters, command.variables, command.ref, ['123', 'css selector', '#body', undefined])
        expect(body).toMatchSnapshot()
    })

    it('should do a proper request with non required params', () => {
        const body = validate(command.command, command.parameters, command.variables, command.ref, ['123', 'css selector', '#body', 123])
        expect(body).toMatchSnapshot()
    })
})

test('getPrototype', () => {
    let i = 0
    expect(getPrototype(() => ++i)).toMatchSnapshot()
})

describe('findElement utils', () => {
    afterEach(() => {
        pageMock.waitForSelector.mockClear()
        pageMock.waitForXPath.mockClear()
        pageMock.$$eval.mockClear()
        pageMock.$.mockClear()
        pageMock.$x.mockClear()
        pageMock.$$.mockClear()
    })

    describe('findElement', () => {
        it('tries to find element using css selector', async () => {
            const scope = {
                timeouts: { get: jest.fn() },
                elementStore: { set: jest.fn().mockReturnValue('foobar') }
            }
            pageMock.$.mockReturnValue(Promise.resolve(42))

            expect(await findElement.call(scope as any, pageMock as any as any, 'css selector', 'barfoo'))
                .toEqual({ 'element-6066-11e4-a52e-4f735466cecf': 'foobar' })

            expect(pageMock.waitForSelector).toBeCalledTimes(0)
            expect(pageMock.$).toBeCalledWith('barfoo')
            expect(pageMock.$x).toBeCalledTimes(0)
        })

        it('tries to find element using xpath', async () => {
            const scope = {
                timeouts: { get: jest.fn() },
                elementStore: { set: jest.fn().mockReturnValue('foobar') }
            }
            pageMock.$x.mockReturnValue(Promise.resolve([42]))

            expect(await findElement.call(scope as any, pageMock as any, 'xpath', '//img'))
                .toEqual({ 'element-6066-11e4-a52e-4f735466cecf': 'foobar' })

            expect(pageMock.waitForSelector).toBeCalledTimes(0)
            expect(pageMock.$).toBeCalledTimes(0)
            expect(pageMock.$x).toBeCalledWith('//img')
        })

        it('should fail if not found', async () => {
            const scope = {
                timeouts: { get: jest.fn() },
                elementStore: { set: jest.fn() }
            }
            pageMock.$.mockReturnValue(Promise.resolve(null))

            const errorMessage = await findElement.call(scope as any, pageMock as any, 'css selector', 'barfoo')
            expect((errorMessage as Error).message)
                .toContain('Element with selector "barfoo" not found')
        })

        it('should not fail with the same error if Puppeteer can not find the element', async () => {
            const scope = {
                timeouts: { get: jest.fn() },
                elementStore: { set: jest.fn() }
            }
            pageMock.$.mockReturnValue(Promise.reject(new Error('failed to find element')))

            const errorMessage = await findElement.call(scope as any, pageMock as any, 'css selector', 'barfoo')
            expect((errorMessage as Error).message)
                .toContain('Element with selector "barfoo" not found')
        })

        it('sets implicit waits', async () => {
            const scope = {
                timeouts: { get: jest.fn().mockReturnValue(1234) },
                elementStore: { set: jest.fn() }
            }
            pageMock.$.mockReturnValue(Promise.resolve('foobar'))
            await findElement.call(scope as any, pageMock as any, 'css selector', 'barfoo')
            expect(pageMock.waitForSelector).toBeCalledTimes(1)
            expect(pageMock.waitForXPath).toBeCalledTimes(0)
        })

        it('sets implicit waits with xpath', async () => {
            const scope = {
                timeouts: { get: jest.fn().mockReturnValue(1234) },
                elementStore: { set: jest.fn() }
            }
            pageMock.$.mockReturnValue(Promise.resolve('foobar'))
            await findElement.call(scope as any, pageMock as any, 'xpath', 'barfoo')
            expect(pageMock.waitForSelector).toBeCalledTimes(0)
            expect(pageMock.waitForXPath).toBeCalledTimes(1)
        })
    })

    describe('findElements', () => {
        it('should find elements', async () => {
            const scope = {
                timeouts: { get: jest.fn() },
                elementStore: { set: jest.fn().mockReturnValue('foobar') }
            }
            pageMock.$$.mockReturnValue(Promise.resolve([42, 11]))
            expect(await findElements.call(scope as any, pageMock as any, 'css selector', 'barfoo')).toMatchSnapshot()
            expect(pageMock.waitForSelector).toBeCalledTimes(0)
            expect(pageMock.$$).toBeCalledWith('barfoo')
            expect(pageMock.$x).toBeCalledTimes(0)
        })

        it('should find elements with xpath', async () => {
            const scope = {
                timeouts: { get: jest.fn() },
                elementStore: { set: jest.fn().mockReturnValue('foobar') }
            }
            pageMock.$x.mockReturnValue(Promise.resolve([42, 11]))
            expect(await findElements.call(scope as any, pageMock as any, 'xpath', 'barfoo')).toMatchSnapshot()
            expect(pageMock.waitForSelector).toBeCalledTimes(0)
            expect(pageMock.$$).toBeCalledTimes(0)
            expect(pageMock.$x).toBeCalledWith('barfoo')
        })

        it('should return immiadiatelly if no elements were found', async () => {
            const scope = {
                timeouts: { get: jest.fn() },
                elementStore: { set: jest.fn() }
            }
            pageMock.$$.mockReturnValue(Promise.resolve([]))
            expect(await findElements.call(scope as any, pageMock as any, 'css selector', 'barfoo')).toEqual([])
            expect(pageMock.waitForSelector).toBeCalledTimes(0)
            expect(scope.elementStore.set).toBeCalledTimes(0)
        })

        it('sets implicit waits', async () => {
            const scope = {
                timeouts: { get: jest.fn().mockReturnValue(1234) },
                elementStore: { set: jest.fn() }
            }
            pageMock.$$.mockReturnValue(Promise.resolve(['foobar']))
            await findElements.call(scope as any, pageMock as any, 'css selector', 'barfoo')
            expect(pageMock.waitForSelector).toBeCalledTimes(1)
        })

        it('sets implicit waits with xpath', async () => {
            const scope = {
                timeouts: { get: jest.fn().mockReturnValue(1234) },
                elementStore: { set: jest.fn() }
            }
            pageMock.$x.mockReturnValue(Promise.resolve(['foobar']))
            await findElements.call(scope as any, pageMock as any, 'xpath', 'barfoo')
            expect(pageMock.waitForSelector).toBeCalledTimes(0)
            expect(pageMock.waitForXPath).toBeCalledTimes(1)
        })
    })
})

describe('sanitizeError', () => {
    it('should set stale element error', () => {
        const err = new Error('Node is detached from document bla bla')
        expect(sanitizeError(err).stack)
            .toContain('stale element reference: The element reference is stale')
    })

    it('should not contain puppeteer traces', () => {
        const err = new Error('some error')
        const newStack = err?.stack?.split('\n') || []
        newStack.push('at /foo/bar/devtools/node_modules/puppeteer-core/bla.js')
        err.stack = newStack.join('\n')
        expect(sanitizeError(err)).not.toContain('devtools/node_modules/puppeteer-core')
    })
})

test('transformExecuteArgs', async () => {
    const scope = { elementStore: new Map() }
    scope.elementStore.set('foobar', 'barfoo')

    expect(await transformExecuteArgs.call(scope as any, [
        'foo',
        { 'element-6066-11e4-a52e-4f735466cecf': 'foobar' },
        true,
        42
    ])).toEqual(['foo', 'barfoo', true, 42])
})

test('transformExecuteArgs throws stale element if element is not in store', async () => {
    const scope = { elementStore: new Map() }
    scope.elementStore.set('foobar', 'barfoo')

    expect(transformExecuteArgs.call(scope as any, [
        'foo',
        { 'element-6066-11e4-a52e-4f735466cecf': 'not-existing' },
        true,
        42
    ])).rejects.toThrow()
})

describe('transformExecuteResult', () => {
    test('multiple results', async () => {
        const scope = {
            timeouts: { get: jest.fn() },
            elementStore: { set: jest.fn().mockReturnValue('foobar') }
        }
        pageMock.$.mockReturnValue(Promise.resolve(42))
        expect(await transformExecuteResult.call(scope as any, pageMock as any, [
            'foobar',
            '__executeElement',
            42
        ])).toMatchSnapshot()
    })

    test('single result', async () => {
        const scope = {
            timeouts: { get: jest.fn() },
            elementStore: { set: jest.fn().mockReturnValue('foobar') }
        }
        pageMock.$.mockReturnValue(Promise.resolve(42))
        expect(await transformExecuteResult.call(scope as any, pageMock as any, 'foobar'))
            .toEqual('foobar')
    })

    afterEach(() => {
        pageMock.waitForSelector.mockClear()
        pageMock.$.mockClear()
        pageMock.$$.mockClear()
        pageMock.$$eval.mockClear()
    })
})

test('getPages', async () => {
    const browser = {
        pages: jest.fn()
            .mockReturnValueOnce([])
            .mockReturnValueOnce([])
            .mockReturnValueOnce([{}])
    }
    await getPages(browser as any, 10)
    expect(browser.pages).toBeCalledTimes(3)
})

test('getStaleElementError', () => {
    const err = getStaleElementError('foobar')
    expect(err instanceof Error).toBe(true)
    expect(err.name).toContain('stale element reference')
})

test('uniq', () => {
    const listA = [1, 2, 3]
    expect(listA).toBe(listA)
    expect(listA).not.toBe(uniq(listA as any))
})

test('findByWhich', () => {
    (canAccess as jest.Mock).mockImplementation(() => true)
    expect(findByWhich(['firefox'], [{ regex: /firefox/, weight: 51 }]))
        .toEqual([])

    ;(childProcess as any).shouldReturn('/path/to/other/firefox\n')
    expect(findByWhich(['firefox'], [{ regex: /firefox/, weight: 51 }]))
        .toEqual(['/path/to/other/firefox'])

    ;(canAccess as jest.Mock).mockImplementation(() => {
        throw new Error('uups')
    })
    expect(findByWhich(['firefox'], [{ regex: /firefox/, weight: 51 }]))
        .toEqual([])
})

test('patchDebug', () => {
    const logMock = { debug: jest.fn() }
    patchDebug(logMock as any)
    debug.log('something something - puppeteer:protocol foobar')
    expect(logMock.debug).toBeCalledWith('foobar')
})

test('patchDebug with debug not install in puppeteer', () => {
    const logMock = { debug: jest.fn() }
    ;(path as any).setResolveResult('pptrDebug')
    patchDebug(logMock as any)
    pptrDebugLog('something something - puppeteer:protocol barfoo')
    expect(logMock.debug).toBeCalledWith('barfoo')
})
