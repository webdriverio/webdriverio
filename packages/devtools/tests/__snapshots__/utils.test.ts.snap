// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findElement utils findElements should find elements 1`] = `
Array [
  Object {
    "element-6066-11e4-a52e-4f735466cecf": "foobar",
  },
  Object {
    "element-6066-11e4-a52e-4f735466cecf": "foobar",
  },
]
`;

exports[`findElement utils findElements should find elements with xpath 1`] = `
Array [
  Object {
    "element-6066-11e4-a52e-4f735466cecf": "foobar",
  },
  Object {
    "element-6066-11e4-a52e-4f735466cecf": "foobar",
  },
]
`;

exports[`getPrototype 1`] = `
Object {
  "acceptAlert": Object {
    "value": 53,
  },
  "addCookie": Object {
    "value": 46,
  },
  "back": Object {
    "value": 8,
  },
  "closeWindow": Object {
    "value": 13,
  },
  "createWindow": Object {
    "value": 15,
  },
  "deleteAllCookies": Object {
    "value": 47,
  },
  "deleteCookie": Object {
    "value": 49,
  },
  "deleteSession": Object {
    "value": 2,
  },
  "dismissAlert": Object {
    "value": 52,
  },
  "elementClear": Object {
    "value": 40,
  },
  "elementClick": Object {
    "value": 39,
  },
  "elementSendKeys": Object {
    "value": 41,
  },
  "executeAsyncScript": Object {
    "value": 44,
  },
  "executeScript": Object {
    "value": 43,
  },
  "findElement": Object {
    "value": 25,
  },
  "findElementFromElement": Object {
    "value": 27,
  },
  "findElements": Object {
    "value": 26,
  },
  "findElementsFromElement": Object {
    "value": 28,
  },
  "forward": Object {
    "value": 9,
  },
  "fullscreenWindow": Object {
    "value": 24,
  },
  "getActiveElement": Object {
    "value": 29,
  },
  "getAlertText": Object {
    "value": 54,
  },
  "getAllCookies": Object {
    "value": 45,
  },
  "getElementAttribute": Object {
    "value": 32,
  },
  "getElementCSSValue": Object {
    "value": 34,
  },
  "getElementComputedLabel": Object {
    "value": 59,
  },
  "getElementComputedRole": Object {
    "value": 58,
  },
  "getElementProperty": Object {
    "value": 33,
  },
  "getElementRect": Object {
    "value": 37,
  },
  "getElementTagName": Object {
    "value": 36,
  },
  "getElementText": Object {
    "value": 35,
  },
  "getNamedCookie": Object {
    "value": 48,
  },
  "getPageSource": Object {
    "value": 42,
  },
  "getTimeouts": Object {
    "value": 4,
  },
  "getTitle": Object {
    "value": 11,
  },
  "getUrl": Object {
    "value": 6,
  },
  "getWindowHandle": Object {
    "value": 12,
  },
  "getWindowHandles": Object {
    "value": 16,
  },
  "getWindowRect": Object {
    "value": 20,
  },
  "isElementDisplayed": Object {
    "value": 31,
  },
  "isElementEnabled": Object {
    "value": 38,
  },
  "isElementSelected": Object {
    "value": 30,
  },
  "maximizeWindow": Object {
    "value": 22,
  },
  "minimizeWindow": Object {
    "value": 23,
  },
  "navigateTo": Object {
    "value": 7,
  },
  "newSession": Object {
    "value": 1,
  },
  "performActions": Object {
    "value": 50,
  },
  "printPage": Object {
    "value": 17,
  },
  "refresh": Object {
    "value": 10,
  },
  "releaseActions": Object {
    "value": 51,
  },
  "sendAlertText": Object {
    "value": 55,
  },
  "setTimeouts": Object {
    "value": 5,
  },
  "setWindowRect": Object {
    "value": 21,
  },
  "status": Object {
    "value": 3,
  },
  "switchToFrame": Object {
    "value": 18,
  },
  "switchToParentFrame": Object {
    "value": 19,
  },
  "switchToWindow": Object {
    "value": 14,
  },
  "takeElementScreenshot": Object {
    "value": 57,
  },
  "takeScreenshot": Object {
    "value": 56,
  },
}
`;

exports[`transformExecuteResult multiple results 1`] = `
Array [
  Array [
    "foobar",
    "__executeElement",
    42,
  ],
  Object {
    "element-6066-11e4-a52e-4f735466cecf": "foobar",
  },
  Array [
    "foobar",
    "__executeElement",
    42,
  ],
]
`;

exports[`validate should do a proper request 1`] = `
Object {
  "elementId": "123",
  "using": "css selector",
  "value": "#body",
}
`;

exports[`validate should do a proper request with non required params 1`] = `
Object {
  "customParam": 123,
  "elementId": "123",
  "using": "css selector",
  "value": "#body",
}
`;
